Source/WebCore/rendering/RenderGrid.cpp

```diff
@@ -778,6 +818,47 @@ void RenderGrid::placeItemsOnGrid(std::optional<LayoutUnit> availableLogicalWidt
 #endif
 }
 
+LayoutUnit RenderGrid::masonryContentSize() const
+{
+    return m_masonryLayout.gridContentSize();
+}
+
+static bool itemGridAreaStartsAtFirstLine(const GridArea& area, GridTrackSizingDirection masonryDirection)
+{
+    return !(masonryDirection == ForRows ? area.rows.startLine() : area.columns.startLine());
+}
+
+bool RenderGrid::hasDefiniteGridAxisPosition(const RenderBox* child, GridTrackSizingDirection gridAxisDirection) const 
+{
+    auto itemSpan = GridPositionsResolver::resolveGridPositionsFromStyle(*this, *child, gridAxisDirection);
+    return !itemSpan.isIndefinite();
+}
+
+static bool itemGridAreaIsWithinImplicitGrid(const GridArea& area, unsigned gridAxisLinesCount, GridTrackSizingDirection gridAxisDirection)
+{
+    auto itemSpan = gridAxisDirection == ForColumns ? area.columns : area.rows;
+    return itemSpan.startLine() <  gridAxisLinesCount && itemSpan.endLine() < gridAxisLinesCount;
+}
+
+void RenderGrid::collectMasonryItems(Grid& grid, unsigned availableGridAxisTracks, GridTrackSizingDirection masonryDirection, HashMap<RenderBox*, GridArea>& itemsOnFirstTrack, Vector<RenderBox*>& itemsWithDefiniteGridAxisPosition, Vector<RenderBox*>& itemsWithIndefinitePosition) const
+{
+    ASSERT(availableGridAxisTracks);
+
+    auto gridAxisDirection = masonryDirection == ForRows ? ForColumns : ForRows;
+    for (auto* child = grid.orderIterator().first(); child; child = grid.orderIterator().next()) {
+        if (grid.orderIterator().shouldSkipChild(*child))
+            continue;
+
+        auto gridArea = grid.gridItemArea(*child);
+        if (itemsOnFirstTrack.size() != availableGridAxisTracks && itemGridAreaStartsAtFirstLine(gridArea, masonryDirection) && itemGridAreaIsWithinImplicitGrid(gridArea, availableGridAxisTracks + 1, gridAxisDirection))
```

Comment:
What is the equivalence between the number of items on the first track and the number of tracks?